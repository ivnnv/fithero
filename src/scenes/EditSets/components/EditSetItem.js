/* @flow */

import * as React from 'react';
import { useCallback, memo } from 'react';
import { StyleSheet, TouchableWithoutFeedback, View } from 'react-native';
import { Text, useTheme } from 'react-native-paper';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

import type { WorkoutSetSchemaType } from '../../../database/types';
import type { ThemeType } from '../../../utils/theme/withTheme';

type Props = {|
  isSelected: boolean,
  trophyColor: ?string,
  color: ?string,
  onPressItem: (setId: string) => void,
  index: number,
  set: WorkoutSetSchemaType,
  children: React.Node,
|};

const EditSetItem = (props: Props) => {
  const { index, isSelected, trophyColor, onPressItem, set, children } = props;

  const theme: ThemeType = useTheme();

  const onPress = useCallback(() => {
    onPressItem(set.id);
  }, [onPressItem, set.id]);

  return (
    <TouchableWithoutFeedback onPress={onPress} testID={`editSetItem-${index}`}>
      <View
        style={[
          styles.item,
          isSelected && { backgroundColor: theme.colors.selected },
        ]}
      >
        <View style={styles.leftContent}>
          <Text style={[styles.text, styles.index]}>{index}.</Text>
          <Icon
            name="trophy"
            size={24}
            color={trophyColor}
            style={[styles.icon, !trophyColor && styles.notMax]}
          />
        </View>
        {children}
      </View>
    </TouchableWithoutFeedback>
  );
};

const styles = StyleSheet.create({
  item: {
    flexDirection: 'row',
    alignItems: 'center',
    height: 48,
    paddingHorizontal: 36,
  },
  text: {
    fontSize: 18,
  },
  leftContent: {
    flexDirection: 'row',
    flex: 0.25,
  },
  index: {
    textAlign: 'left',
  },
  icon: {
    marginHorizontal: 24,
    height: 24,
    width: 24,
  },
  notMax: {
    opacity: 0,
  },
});

// TODO
// $FlowFixMe
export default memo(EditSetItem, (prevProps, nextProps) => {
  if (
    prevProps.isSelected !== nextProps.isSelected ||
    prevProps.trophyColor !== nextProps.trophyColor ||
    // TODO comment out for now because it breaks as selectedId always changes onPressItem.
    //  It seems to not affect the usage (as already we have isSelected prop)
    // prevProps.onPressItem !== nextProps.onPressItem ||
    prevProps.index !== nextProps.index
  ) {
    return false;
  }

  return (
    // Doing it like this because of Realm
    JSON.stringify(prevProps.set) === JSON.stringify(nextProps.set)
  );
});
