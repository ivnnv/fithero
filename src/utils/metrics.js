/* @flow */

import * as RNLocalize from 'react-native-localize';

import type { WorkoutExerciseWeightRepsType } from '../database/types';
import type { DefaultUnitSystemType } from '../redux/modules/settings';

const oneKgToLb = 2.204622622;

export const toLb = (value: number) => value * oneKgToLb;
export const toKg = (value: number) => value / oneKgToLb;
export const toTwoDecimals = (value: number) => Math.round(value * 100) / 100;

export const getWeight = (
  value: number,
  exercise: ?WorkoutExerciseWeightRepsType,
  defaultUnitSystem: DefaultUnitSystemType
) => {
  const unit =
    exercise &&
    ((exercise.isValid && exercise.isValid()) || !exercise.isValid) &&
    exercise.weight_unit
      ? exercise.weight_unit
      : defaultUnitSystem;
  if (unit !== 'metric') {
    return toLb(value);
  }
  return value;
};

export const getWeightUnit = (
  exercise: ?WorkoutExerciseWeightRepsType,
  defaultUnitSystem: DefaultUnitSystemType
): DefaultUnitSystemType =>
  exercise &&
  ((exercise.isValid && exercise.isValid()) || !exercise.isValid) &&
  exercise.weight_unit
    ? exercise.weight_unit
    : defaultUnitSystem;

export const getDefaultUnitSystemByCountry = (): DefaultUnitSystemType => {
  const country = RNLocalize.getCountry();
  if (
    country === 'US' ||
    country === 'UK' ||
    country === 'MM' ||
    country === 'LR'
  ) {
    return 'imperial';
  }
  return 'metric';
};
