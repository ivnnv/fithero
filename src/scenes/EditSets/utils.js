/* @flow */

import { getLastSetByType } from '../../database/services/WorkoutSetService';
import type {
  WorkoutExerciseWeightRepsType,
  WorkoutSetWeightRepsType,
} from '../../database/types';
import { getWeight, toTwoDecimals } from '../../utils/metrics';
import type { DefaultUnitSystemType } from '../../redux/modules/settings';

export const getLastSet = (
  exercise: ?WorkoutExerciseWeightRepsType,
  exerciseKey: string
) => {
  if (exercise) {
    return exercise.sets[exercise.sets.length - 1];
  } else {
    const sets = getLastSetByType(exerciseKey || exercise?.type);
    if (sets.length > 0) {
      return sets[0];
    }

    return null;
  }
};

export const getLastWeight = (
  exercise: ?WorkoutExerciseWeightRepsType,
  lastSet: ?WorkoutSetWeightRepsType,
  unit: DefaultUnitSystemType
) => {
  const defaultWeight = unit === 'metric' ? 20 : 45;

  const lastWeight = lastSet
    ? toTwoDecimals(getWeight(lastSet.weight, exercise, unit))
    : defaultWeight;

  return lastWeight.toString();
};
