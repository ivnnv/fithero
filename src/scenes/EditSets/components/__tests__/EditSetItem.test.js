/* @flow */

import React from 'react';
import { render } from 'react-native-testing-library';
import { Provider } from 'react-native-paper';

import EditSetItem from '../EditSetItem';
import { toDate } from '../../../../utils/date';
import { defaultTheme } from '../../../../utils/theme';

describe('EditSetItem', () => {
  const _render = props =>
    render(
      <Provider theme={defaultTheme}>
        <EditSetItem
          isSelected={false}
          onPressItem={jest.fn()}
          index={1}
          set={{
            id: '2018-05-04T00:00:00.000Z_bench-press_001',
            reps: 7,
            weight: 70,
            date: toDate('2018-05-04T00:00:00.000Z'),
            type: 'bench-press',
          }}
          trophyColor={null}
          // $FlowFixMe
          {...props}
        />
      </Provider>
    );

  it('shows the set selected if is set is isSelected', () => {
    const { toJSON: toJSON } = _render();
    const { toJSON: toJSONSelected } = _render({ isSelected: true });

    // $FlowFixMe
    expect(toJSON()).toMatchDiffSnapshot(toJSONSelected(), {
      contextLines: 0,
      stablePatchmarks: true,
    });
  });

  it('shows a trophy if set is maxSetId', () => {
    const { toJSON: toJSON } = _render();
    const { toJSON: toJSONMax } = _render({ trophyColor: 'yellow' });

    // $FlowFixMe
    expect(toJSON()).toMatchDiffSnapshot(toJSONMax(), {
      contextLines: 0,
      stablePatchmarks: true,
    });
  });
});
