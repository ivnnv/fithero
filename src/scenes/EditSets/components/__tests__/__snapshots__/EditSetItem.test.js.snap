// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSetItem shows a trophy if set is maxSetId 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ --- --- @@
-               "color": null,
+               "color": "yellow",
@@ --- --- @@
-               Object {
-                 "opacity": 0,
-               },
+               false,
`;

exports[`EditSetItem shows the set selected if is set is isSelected 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ --- --- @@
-         false,
+         Object {
+           "backgroundColor": "#66809B",
+         },
`;
